### ISSUE TITLE
你这么写训练内存会崩的。

### ISSUE THREAD

**Issue by liangtianxin** [ByronAndrade/moire_detection#1]
`#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Aug 11 18:18:04 2022
"""

from matplotlib import pyplot as plt
import numpy as np
import sys
import argparse
from os import listdir
from os.path import isfile, join
from PIL import Image
from sklearn import preprocessing
from skimage import io
from sklearn.model_selection import train_test_split
import os
from mCNN import createModel
from keras.utils import np_utils # utilities for one-hot encoding of ground truth values
from keras.callbacks import ModelCheckpoint

from tqdm import tqdm

#constants
WIDTH = 500#384
HEIGHT = 375#512

#constants
WIDTH = 500  #384
HEIGHT = 375  #512
#constants
width = 500   #384 
height = 375  #512 
depth = 1
num_classes = 2

def scaleData(inp, minimum, maximum):
    minMaxScaler = preprocessing.MinMaxScaler(copy=True, feature_range=(minimum,maximum))
    inp = inp.reshape(-1, 1)
    inp = minMaxScaler.fit_transform(inp)
    
    return inp

def readAndScaleImage_new(f, customStr, trainImagePath, label):
    
    imgVector_LL,imgVector_LH,imgVector_HL,imgVector_HH = list(),list(),list(),list()
    
    label_cluster = list()
    
    fileName = (os.path.splitext(f)[0])
    fLL = (f.replace(fileName, fileName + customStr + '_LL')).replace('.jpg','.tiff')
    fLH = (f.replace(fileName, fileName + customStr + '_LH')).replace('.jpg','.tiff')
    fHL = (f.replace(fileName, fileName + customStr + '_HL')).replace('.jpg','.tiff')
    fHH = (f.replace(fileName, fileName + customStr + '_HH')).replace('.jpg','.tiff')
    
    try:
        imgLL = Image.open(join(trainImagePath, fLL))
        imgLH = Image.open(join(trainImagePath, fLH))
        imgHL = Image.open(join(trainImagePath, fHL))
        imgHH = Image.open(join(trainImagePath, fHH))
    except Exception as e:
        print('Error: Couldnt read the file {}. Make sure only images are present in the folder'.format(fileName))
        print('Exception:', e)
        return None
        
    imgLL = np.array(imgLL)
    imgLH = np.array(imgLH)
    imgHL = np.array(imgHL)
    imgHH = np.array(imgHH)
    imgLL = scaleData(imgLL, 0, 1)
    imgLH = scaleData(imgLH, -1, 1)
    imgHL = scaleData(imgHL, -1, 1)
    imgHH = scaleData(imgHH, -1, 1)
        
    
    imgVector = imgLL.reshape(1, WIDTH*HEIGHT)
    imgVector_LL.append(imgVector)
    imgVector = imgLH.reshape(1, WIDTH*HEIGHT)
    imgVector_LH.append(imgVector)
    imgVector = imgHL.reshape(1, WIDTH*HEIGHT)
    imgVector_HL.append(imgVector)
    imgVector = imgHH.reshape(1, WIDTH*HEIGHT)
    imgVector_HH.append(imgVector)
    
    label_cluster.append(label)  
    
    return imgVector_LL,imgVector_LH,imgVector_HL,imgVector_HH,label_cluster

def readImageSet_new(positiveImagePath, trainImagePath,label):
    
    imgVector_LL_ALL = list()
    imgVector_LH_ALL = list()
    imgVector_HL_ALL = list()
    imgVector_HH_ALL = list()
    label_cluster_ALL = list()
    
    for root, dirs, files in os.walk(positiveImagePath, topdown=False):
        for f in tqdm(files):
            if f.split('.')[1] == 'jpg':
                imgVector_LL_,imgVector_LH_,imgVector_HL_,imgVector_HH_,label_cluster_ = readAndScaleImage_new(f, '', trainImagePath, label)
                
                imgVector_LL_180,imgVector_LH_180,imgVector_HL_180,imgVector_HH_180,label_cluster_180 = readAndScaleImage_new(f, '_180', trainImagePath, label)
                
                imgVector_LL_180_FLIP,imgVector_LH_180_FLIP,imgVector_HL_180_FLIP,imgVector_HH_180_FLIP,label_cluster_180_FLIP=  readAndScaleImage_new(f, '_180_FLIP', trainImagePath, label)
            
                # 合并后继续累加，保证返回的完整的一列
                imgVector_LL = imgVector_LL_ + imgVector_LL_180 + imgVector_LL_180_FLIP
                imgVector_LL_ALL = imgVector_LL_ALL +imgVector_LL
                
                imgVector_LH = imgVector_LH_ + imgVector_LH_180 + imgVector_LH_180_FLIP
                imgVector_LH_ALL = imgVector_LH_ALL +imgVector_LH
                
                imgVector_HL = imgVector_HL_ + imgVector_HL_180 + imgVector_HL_180_FLIP
                imgVector_HL_ALL = imgVector_HL_ALL + imgVector_HL
                
                imgVector_HH = imgVector_HH_ + imgVector_HH_180 + imgVector_HH_180_FLIP
                imgVector_HH_ALL = imgVector_HH_ALL + imgVector_HH
                
                label_cluster = label_cluster_ + label_cluster_180 +label_cluster_180_FLIP
                label_cluster_ALL = label_cluster_ALL + label_cluster
            else:
                print ('find image is not jpg!')
    
    return imgVector_LL_ALL,imgVector_HL_ALL,imgVector_HL_ALL,imgVector_HH_ALL,label_cluster_ALL

# 我这个写法下就没有第四列为空的情况。

positiveImagePath = 'positiveImages'

positiveTrainImagePath = 'trainDataPositive'

pos_label = 0

imgVector_LL_pos,imgVector_HL_pos,imgVector_HL_pos,imgVector_HH_pos,label_cluster_pos= readImageSet_new(positiveImagePath, positiveTrainImagePath,pos_label)

negativeImagePath = 'negativeImages'

negativeTrainImagePath = 'trainDataNegative'

neg_label = 1

imgVector_LL_neg,imgVector_HL_neg,imgVector_HL_neg,imgVector_HH_neg,label_cluster_neg= readImageSet_new(negativeImagePath, negativeTrainImagePath,neg_label)

train_data_list = list()

for imgVector_LL,imgVector_HL,imgVector_HL,imgVector_HH,label_cluster in zip(imgVector_LL_pos,imgVector_HL_pos,imgVector_HL_pos,imgVector_HH_pos,label_cluster_pos):
    train_data_list.append((imgVector_LL,imgVector_HL,imgVector_HL,imgVector_HH,label_cluster))

for imgVector_LL,imgVector_HL,imgVector_HL,imgVector_HH,label_cluster in zip(imgVector_LL_neg,imgVector_HL_neg,imgVector_HL_neg,imgVector_HH_neg,label_cluster_neg):
    train_data_list.append((imgVector_LL,imgVector_HL,imgVector_HL,imgVector_HH,label_cluster))

print ('全部数据能够载入，不会崩掉')
print (len(train_data_list))

# 数据集合划分
mode = 0
random_order = np.arange(len(train_data_list))
train_list = [train_data_list[j] for i, j in enumerate(random_order) if i % 9 != mode]
vad_list = [train_data_list[j] for i, j in enumerate(random_order) if i % 9 == mode]
print(len(train_list), len(vad_list))

class train_data_generator:
    def __init__(self, train_list, batch_size):
        self.height = 375
        self.width = 500
        self.depth = 1
        self.train_list = train_list
        self.batch_size = batch_size
        self.steps = len(self.train_list) // self.batch_size
        if len(self.train_list) % self.batch_size != 0:
            self.steps += 1
            
    def __len__(self):
        return self.steps
    
    def __iter__(self):
        while True:
            # 返回训练数据集中的索引列表
            idxs = np.arange(len(self.train_list))
            np.random.shuffle(idxs) #乱序
            X_LL_ALL,X_LH_ALL,X_HL_ALL,X_HH_ALL,y_all = list(),list(),list(),list(),list()
            for i in idxs: # 这些仅仅循环一个batch size
                train = self.train_list[i]
                
                X_LL_train = train[0] 
                X_LL_ALL.append(X_LL_train)
                
                X_LH_train = train[1] 
                X_LH_ALL.append(X_LH_train)

                X_HL_train = train[2] 
                X_HL_ALL.append(X_HL_train)

                X_HH_train = train[3] 
                X_HH_ALL.append(X_HH_train)

                y_train = train[-1]
                y_single = np_utils.to_categorical(y_train, 2) # 测试0 和1 后效果可以是onehot
                y_all.append(y_single)
                
                if len(X_LL_ALL) == self.batch_size or i == idxs[-1]: # 要不断增加 X_LL_ALL
                    
                    X_LL_ALL = np.array(X_LL_ALL)
                    X_LL_ALL = np.squeeze(X_LL_ALL, axis=1)
                    
                    X_LH_ALL = np.array(X_LH_ALL)
                    X_LH_ALL = np.squeeze(X_LH_ALL, axis=1)
                    
                    X_HL_ALL = np.array(X_HL_ALL)
                    X_HL_ALL = np.squeeze(X_HL_ALL, axis=1)
                    
                    X_HH_ALL = np.array(X_HH_ALL)
                    X_HH_ALL = np.squeeze(X_HH_ALL, axis=1)
                    
                    y_all = np.array(y_all)
                    #y_all = np.squeeze(y_all, axis=1)
                    if len(X_LL_ALL) == batch_size:
                        X_LL_ALL = X_LL_ALL.reshape((self.batch_size, self.height, self.width, self.depth))
                        X_LH_ALL = X_LH_ALL.reshape((self.batch_size, self.height, self.width, self.depth))
                        X_HL_ALL = X_HL_ALL.reshape((self.batch_size, self.height, self.width, self.depth))
                        X_HH_ALL = X_HH_ALL.reshape((self.batch_size, self.height, self.width, self.depth))
                    else:
                        X_LL_ALL = X_LL_ALL.reshape((len(X_LL_ALL), self.height, self.width, self.depth))
                        X_LH_ALL = X_LH_ALL.reshape((len(X_LL_ALL), self.height, self.width, self.depth))
                        X_HL_ALL = X_HL_ALL.reshape((len(X_LL_ALL), self.height, self.width, self.depth))
                        X_HH_ALL = X_HH_ALL.reshape((len(X_LL_ALL), self.height, self.width, self.depth))

                    yield [X_LL_ALL, X_LH_ALL, X_HL_ALL, X_HH_ALL], y_all  # 输入和输出用分号隔开；
                    X_LL_ALL,X_LH_ALL,X_HL_ALL,X_HH_ALL,y_all = list(),list(),list(),list(),list()

checkPointFolder = 'checkPoint'
checkpoint_name = checkPointFolder + '/Weights-{epoch:03d}--{val_loss:.5f}.hdf5' 
checkpoint = ModelCheckpoint(checkpoint_name, monitor='val_loss', verbose = 1, save_best_only = True, mode ='auto')
callbacks_list = [checkpoint]

if not os.path.exists(checkPointFolder):
    os.makedirs(checkPointFolder)

batch_size = 32
epochs = 2
NUM_TRAIN = len(train_list)
NUM_TEST = len(vad_list)
train_generator = train_data_generator(train_list,batch_size)
validation_generator = train_data_generator(vad_list,batch_size)

#print (X_LL_train.shape) #(2178, 375, 500, 1)
# 输入的数据为 这个形式，
cnn_model = createModel(height, width, depth, num_classes)

cnn_model.compile(loss='categorical_crossentropy', # using the cross-entropy loss function
              optimizer='adam', # using the Adam optimiser
              metrics=['accuracy']) # reporting the accuracy

cnn_model.fit_generator(train_generator.__iter__(),
    steps_per_epoch=len(train_generator),
    epochs=epochs,
    validation_data=validation_generator.__iter__(),
    validation_steps= NUM_TEST //batch_size,
    verbose=1,
    use_multiprocessing=True,
    workers=1,
    callbacks=callbacks_list
)

cnn_model.save('moirePattern3CNN_tst01.h5') #保存成功，callbacks 成功

`
